puma/hlplots/tagger.py

```diff
@@ -0,0 +1,214 @@
+"""Tagger module for high level API."""
+import h5py
+import numpy as np
+import pandas as pd
+from numpy.lib.recfunctions import structured_to_unstructured
+
+from puma.utils import calc_disc, logger
+
+
+class TaggerBase:  # pylint: disable=too-many-instance-attributes
+    """Base class handling tagger results."""
+
+    def __init__(self, model_name: str) -> None:
+        """Init Tagger class.
+
+        Parameters
+        ----------
+        model_name : str
+            Name of the model, also correspondinng to the pre-fix of the tagger
+            variables.
+        """
+        self.model_name = model_name
+        self.label = None
+        self.reference = False
+
+        self.scores = None
+        self.perf_var = None
+        self.flvs = ["pu", "pc", "pb"]
+
+        self.is_b = None
+        self.is_light = None
+        self.is_c = None
+
+    def extract_tagger_scores(
+        self, source: object, source_type: str = "data_frame", key: str = None
+    ):
+        """Extract tagger scores from data frame or file.
+
+        Parameters
+        ----------
+        source : object
+            pd.DataFrame or file path to h5 file containing pd.DataFrame or structured
+            numpy array
+        source_type : str, optional
+            Indicates from which source scores should be extracted. Possible options are
+            `data_frame` when passing a pd.DataFrame, `data_frame_path` when passing a
+            file path to a h5 file with a pd.DataFrame or `numpy_structured` when
+            passing a file path to a h5 file with a structured numpy array,
+            by default "data_frame"
+        key : str, optional
+            Key within h5 file, needs to be provided when using the `source_type`
+            `data_frame_path` or `numpy_structured`, by default None
+        Raises
+        ------
+        ValueError
+            if source_type is wrongly specified
+        """
+        # list tagger variables
+        tagger_vars = [f"{self.model_name}_{flv}" for flv in self.flvs]
+        # TODO: change to cae syntax in python 3.10
+        if source_type == "data_frame":
+            logger.debug("Retrieving tagger `%s` from data frame.", self.model_name)
+            self.scores = source[tagger_vars].values
+            return
+        if key is None:
+            raise ValueError(
+                "When using a `source_type` other than `data_frame`, you need to "
+                "specify the `key`."
+            )
+        if source_type == "data_frame_path":
+            logger.debug(
+                "Retrieving tagger %s in data frame from file %s.",
+                self.model_name,
+                source,
+            )
+            df_in = pd.read_hdf(source, key=key)
+            self.scores = df_in[tagger_vars].values
```

Comment:
I would suggest that `self.scores` could be a dataframe with columns for each output probability so as not to have any ambiguity in the columns.

Reply:
`self.scores` are mostly not touched by the user when using `extract_tagger_scores`, so my motivation was that one could directly pipe in the results from a pytorch/keras tagger evaluation which is in exactly this format since that is the only use case where you would need to set `scores` directly.
Together with `self.flvs` which is also used then in other places, e.g. to calculate discriminant etc. this is internally a kind of data frame in this sense.
If you think that's not reasonable, I can also change that.

Reply:
moving it then also in a follow-up issue